shared_model/validators/transaction_batch_validator.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "validators/transaction_batch_validator.hpp"
+
+#include <boost/range/adaptor/indirected.hpp>
+#include <boost/range/combine.hpp>
+#include "interfaces/iroha_internal/batch_meta.hpp"
+#include "interfaces/transaction.hpp"
+
+namespace {
+  enum class BatchCheckResult {
+    kOk,
+    kNoBatchMeta,
+    kIncorrectBatchMetaSize,
+    kIncorrectHashes
+  };
+  /**
+   * Check that all transactions from the collection are mentioned in batch_meta
+   * and are positioned correctly
+   * @param transactions to be checked
+   * @return enum, reporting about success result or containing a found error
+   */
+  BatchCheckResult batchIsWellFormed(
+      const shared_model::interface::types::TransactionsForwardCollectionType
+          &transactions) {
+    auto batch_meta_opt = transactions.begin()->batchMeta();
+    if (not batch_meta_opt and boost::size(transactions) == 1) {
+      // batch is created from one tx - there is no batch_meta in valid case
+      return BatchCheckResult::kOk;
+    }
+    if (not batch_meta_opt) {
+      // in all other cases batch_meta must present
+      return BatchCheckResult::kNoBatchMeta;
+    }
+
+    const auto &batch_hashes = batch_meta_opt->get()->reducedHashes();
+    if (batch_hashes.size() != boost::size(transactions)) {
```

Comment:
boost::size might have linear complexity for some ranges. Consider saving it to a variable and reusing it for all checks.